File: backend/src/main/java/com/skewwhiffy/auraltester/notation/model/note/AbsoluteNote.java
package com.skewwhiffy.auraltester.notation.model.note;

import com.skewwhiffy.auraltester.helper.NoParallelStream;
import com.skewwhiffy.auraltester.notation.model.interval.DirectedInterval;
import com.skewwhiffy.auraltester.notation.model.interval.Interval;
import com.skewwhiffy.auraltester.notation.model.key.Key;
import lombok.val;
import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public record AbsoluteNote(Note note, Octave octave, Optional<String> lyric) implements Comparable<AbsoluteNote> {
    public AbsoluteNote apply(DirectedInterval interval) {
        return switch (interval.direction()) {
            case UP -> plus(interval.interval());
            case DOWN -> minus(interval.interval());
        };
    }

    public AbsoluteNote withLyric(String lyric) {
        return new AbsoluteNote(note, octave, Optional.of(lyric));
    }

    public AbsoluteNote plus(Interval interval) {
        val defaultNote = switch (interval.degree()) {
            case 1 -> this;
            case 2 -> upMajorSecond();
            case 3 -> upMajorSecond().upMajorSecond();
            case 4 -> plus(Interval.major(3)).upMinorSecond();
            case 5 -> plus(Interval.perfect(4)).upMajorSecond();
            case 6 -> plus(Interval.perfect(5)).upMajorSecond();
            case 7 -> plus(Interval.major(6)).upMajorSecond();
            case 8 -> new AbsoluteNote(note, octave.up(), lyric);
            default -> throw new IllegalArgumentException(
                    MessageFormat.format(
                            "Interval degree of ${interval.degree} not supported",
                            interval.degree()
                    )
            );
        };

        val deviation = interval.deviation();
        if (deviation < 0) {
            return IntStream
                    .range(0, -deviation)
                    .boxed()
                    .reduce(defaultNote, (note, it) -> note.flatten(), NoParallelStream.get());
        }
        if (deviation > 0) {
            return IntStream
                    .range(0, deviation)
                    .boxed()
                    .reduce(defaultNote, (note, it) -> note.sharpen(), NoParallelStream.get());
        }
        return defaultNote;
    }

    public AbsoluteNote minus(Interval interval) {
        val defaultNote = switch (interval.degree()) {
            case 1 -> this;
            case 2 -> downMajorSecond();
            case 3 -> downMajorSecond().downMajorSecond();
            case 4 -> minus(Interval.major(3)).downMinorSecond();
            case 5 -> minus(Interval.perfect(4)).downMajorSecond();
            case 6 -> minus(Interval.perfect(5)).downMajorSecond();
            case 7 -> minus(Interval.major(6)).downMajorSecond();
            case 8 -> new AbsoluteNote(note, octave.down(), lyric);
            default -> throw new IllegalArgumentException();
        };
        val deviation = interval.deviation();
        if (deviation < 0) {
            return IntStream
                    .range(0, -deviation)
                    .boxed()
                    .reduce(
                            defaultNote,
                            (it, i) -> it.sharpen(),
                            NoParallelStream.get()
                    );
        }
        if (deviation > 0) {
            return IntStream
                    .range(0, deviation)
                    .boxed()
                    .reduce(
                            defaultNote,
                            (it, i) -> it.flatten(),
                            NoParallelStream.get()
                    );
        }
        return defaultNote;
    }

    public String getAbc(Key key) {
        return key.getAbc(this);
    }

    public String getWordAbc() {
        return lyric.orElse("*");
    }

    public AbsoluteNote downOne() {
        return minus(Interval.minor(2)).ignoreAccidental();
    }

    public AbsoluteNote upOne() {
        return plus(Interval.minor(2)).ignoreAccidental();
    }

    public AbsoluteNote skipOne() {
        return this.plus(Interval.minor(3)).ignoreAccidental();
    }

    public AbsoluteNote ignoreAccidental() {
        val newNote = new Note(note.noteName(), Accidental.getNatural());
        return new AbsoluteNote(newNote, octave, lyric);
    }

    public AbsoluteNote withNoteName() {
        return withLyric(note.getDisplayString());
        //get() = withLyric(note.displayString)
    }

    private AbsoluteNote sharpen() {
        return new AbsoluteNote(note.sharpen(), octave, lyric);
    }

    private AbsoluteNote flatten() {
        return new AbsoluteNote(note.flatten(), octave, lyric);
    }

    private AbsoluteNote upMajorSecond() {
        return new AbsoluteNote(
                note.upMajorSecond(),
                "B".equals(note.noteName()) ? octave.up() : octave,
                lyric
        );
    }

    private AbsoluteNote upMinorSecond() {
        val upMajorSecond = upMajorSecond();
        return new AbsoluteNote(
                upMajorSecond.note.flatten(),
                upMajorSecond.octave,
                lyric
        );
    }

    private AbsoluteNote downMajorSecond() {
        return new AbsoluteNote(
                note.downMajorSecond(),
                "C".equals(note.noteName()) ? octave.down() : octave,
                lyric
        );
    }

    private AbsoluteNote downMinorSecond() {
        val downMajorSecond = downMajorSecond();
        return new AbsoluteNote(downMajorSecond.note.sharpen(), downMajorSecond.octave, lyric);
    }

    public static AbsoluteNote getMiddleC() {
        return new AbsoluteNote(Note.getC(), Octave.getDefault(), Optional.empty());
    }

    @Override
    public int compareTo(@NotNull AbsoluteNote o) {
        if (this == o) {
            return 0;
        }
        if (octave.equals(o.octave)) {
            return note.compareTo(o.note);
        }
        return octave.compareTo(o.octave);
    }

    @Override
    public String toString() {
        return getAbc(Key.getCMajor());
    }

    public static List<AbsoluteNote> range(AbsoluteNote lower, AbsoluteNote upper) {
        val nonNaturalNotes = Stream
                .of(lower, upper)
                .filter(it -> it.note.accidental().offset() != 0)
                .toList();
        if (!nonNaturalNotes.isEmpty()) {
            throw new IllegalArgumentException(
                    MessageFormat.format(
                            "Note(s) '{0}' have accidentals.",
                            nonNaturalNotes.stream().map(it -> it.getAbc(Key.getCMajor())).collect(Collectors.joining(", "))
                    )
            );
        }
        return lower.compareTo(upper) > 0
                ? range(upper, lower)
                : lower.equals(upper)
                ? Collections.singletonList(lower)
                : Stream.concat(Stream.of(lower), range(lower.upOne(), upper).stream()).toList();
    }

}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/notation/model/note/AbsoluteNote.kt
package com.skewwhiffy.auraltester.notation.model.note

import com.skewwhiffy.auraltester.dao.AbsoluteNoteDao
import com.skewwhiffy.auraltester.notation.model.abc.upOne
import com.skewwhiffy.auraltester.notation.model.interval.DirectedInterval
import com.skewwhiffy.auraltester.notation.model.interval.Interval
import com.skewwhiffy.auraltester.notation.model.interval.IntervalDirection
import com.skewwhiffy.auraltester.notation.model.key.Key

data class AbsoluteNote(val note: Note, val octave: Octave, val lyric: String?) : Comparable<AbsoluteNote> {
    constructor(note: Note, octave: Octave) : this(note, octave, null)

    companion object {
        val middleC: AbsoluteNote
            get() = AbsoluteNote(Note.c, Octave.default, null)

        fun range(lower: AbsoluteNote, upper: AbsoluteNote): List<AbsoluteNote> {
            val nonNaturalNotes = listOf(lower, upper).filter { it.note.accidental.offset != 0 }
            if (nonNaturalNotes.isNotEmpty()) {
                throw IllegalArgumentException(
                    "Note(s) '${nonNaturalNotes.joinToString(", ") { it.getAbc(Key.cMajor) }}' have accidentals.",
                )
            }
            return if (lower > upper) range(upper, lower)
            else if (lower == upper) listOf(lower)
            else listOf(lower) + range(lower.upOne, upper)
        }
    }

    val dao by lazy { AbsoluteNoteDao(note.dao, octave.dao, lyric) }

    fun apply(interval: DirectedInterval) = when (interval.direction) {
        IntervalDirection.UP -> this + interval.interval
        IntervalDirection.DOWN -> this - interval.interval
    }

    fun withLyric(lyric: String): AbsoluteNote = AbsoluteNote(note, octave, lyric)

    operator fun plus(interval: Interval): AbsoluteNote {
        val defaultNote: AbsoluteNote = when (interval.degree) {
            1 -> this
            2 -> upMajorSecond
            3 -> upMajorSecond.upMajorSecond
            4 -> (this + Interval.major(3)).upMinorSecond
            5 -> (this + Interval.perfect(4)).upMajorSecond
            6 -> (this + Interval.perfect(5)).upMajorSecond
            7 -> (this + Interval.major(6)).upMajorSecond
            8 -> AbsoluteNote(note, octave.up, lyric)
            else -> throw IllegalArgumentException("Interval degree of ${interval.degree} not supported")
        }

        return interval.deviation.let {
            when {
                it > 0 -> (1..it).fold(defaultNote) { curr, _ -> curr.sharpen }
                else -> (1..-it).fold(defaultNote) { curr, _ -> curr.flatten }
            }
        }
    }

    operator fun minus(interval: Interval): AbsoluteNote {
        if (interval.degree > 8) {
            val intervalLessAnOctave = Interval(interval.degree - 7, interval.deviation)
            return this - Interval.octave - intervalLessAnOctave
        }
        val defaultNote = when (interval.degree) {
            1 -> this
            2 -> downMajorSecond
            3 -> downMajorSecond.downMajorSecond
            4 -> (this - Interval.major(3)).downMinorSecond
            5 -> (this - Interval.perfect(4)).downMajorSecond
            6 -> (this - Interval.perfect(5)).downMajorSecond
            7 -> (this - Interval.major(6)).downMajorSecond
            8 -> AbsoluteNote(note, octave.down, lyric)
            else -> throw IllegalArgumentException()
        }
        return interval.deviation.let {
            when {
                it > 0 -> (1..it).fold(defaultNote) { curr, _ -> curr.flatten }
                else -> (1..-it).fold(defaultNote) { curr, _ -> curr.sharpen }
            }
        }
    }

    operator fun minus(other: AbsoluteNote): DirectedInterval {
        if (other < this) {
            return -(other - this)
        }
        val intervalNumber = (generateSequence(1) { it + 1 })
            .first { this + Interval(it, 0) >= other }
        val intervalDeviation = (generateSequence(0) { it + 1 })
            .flatMap { listOf(it, -it) }
            .first { this + Interval(intervalNumber, it) == other }
        return Interval(intervalNumber, intervalDeviation)
            .let { DirectedInterval(it, IntervalDirection.DOWN) }
    }

    fun getAbc(key: Key): String {
        return key.getAbc(this)
    }

    val wordAbc: String
        get() = lyric ?: "*"

    private val sharpen: AbsoluteNote
        get() = AbsoluteNote(note.sharpen, octave, lyric)

    private val flatten: AbsoluteNote
        get() = AbsoluteNote(note.flatten, octave, lyric)

    private val upMajorSecond: AbsoluteNote
        get() = AbsoluteNote(
            note.upMajorSecond,
            if ("B" == note.noteName) octave.up else octave,
            lyric
        )

    private val upMinorSecond: AbsoluteNote
        get() = AbsoluteNote(
            upMajorSecond.note.flatten,
            upMajorSecond.octave,
            lyric
        )

    private val downMajorSecond: AbsoluteNote
        get() = AbsoluteNote(
            note.downMajorSecond,
            if (note.noteName == "C") octave.down else octave,
            lyric
        )

    private val downMinorSecond: AbsoluteNote
        get() = AbsoluteNote(
            downMajorSecond.note.sharpen,
            downMajorSecond.octave,
            lyric
        )

    override fun compareTo(other: AbsoluteNote): Int {
        if (this == other) {
            return 0
        }
        if (octave == other.octave) {
            return note.compareTo(other.note)
        }
        return octave.compareTo(other.octave)
    }

    override fun toString() = getAbc(Key.cMajor)
}
